Support a group decomposed search space and apply it to TPE

HideakiImamura opened the pull request:
## Motivation
In the current Optuna, `sample_relative` allows sampling considering the correlation between variables. However, only `IntersectionSearchSpace` is supported as a search space for this purpose. In this PR, the `group` option for `TPESampler` is newly supported as an option that allows sampling considering the correlation between variables in each branch even if the search space is branched by the progress of the trials.

## Description of the changes
- Add `group` option to `TPESampler`.
- Fix directory structure of search spaces.
- Fix some modules to support the `group` option.

## TODOs
- [x] Implementation the feature
- [x] Tests
- [x] Docs

HideakiImamura commented:
@hvy @not522 This PR is ready for review. PTAL.

hvy commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,103 @@
+    return group
+
+
+class _GroupDecomposedSearchSpace(object):
```
Mostly curious but is an instance of this class intended to be used across trials (like the `IntersectionSearchSpace`)? I don't think that the cursor is needed in that case which would simplify the logic.

HideakiImamura replied:
Thanks for the insightful comment. The `_GroupDecomposedSearchSpace` class is instantiated in every call of `infer_relative_search_space`, so we can remove the `cursor`. I'll simplify the logic.

HideakiImamura replied:
I made TPESampler own the `_GroupDecomposedSearchSpace` so that it doesn't have to check every trial in each `infer_relatice_search_space` call. To achieve this, I decided to use the cursor again.

hvy commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,103 @@
+            if not trial.state.is_finished():
+                next_cursor = trial.number
+
+            if trial.state not in states_of_interest:
```
How about filtering the states in `get_trials`, to simplify the logic a bit and to speed things up?

hvy commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,103 @@
+        return group
+
+    for search_space in group:
+        _keys = set(search_space.keys())
```
I don't see this a lot in Optuna but is there a specific meaning to prefixing local variables with underscores?

HideakiImamura replied:
Sorry for the confusion. There is no specific meaning to do that. I'll use the conventional variable names, such as `keys, dist_keys, intersection`.

hvy commented on tests/samplers_tests/search_space_tests/test_group_decomposed.py:
```diff
@@ -0,0 +1,214 @@
+from optuna.trial import Trial
+
+
+def test_search_space_group() -> None:
```
Pedagogic tests 👍

HideakiImamura commented:
@hvy Thanks for the review. I addressed all of your reviews. PTAL.

HideakiImamura commented:
I updated the title and the description of this PR.

HideakiImamura commented:
Let me leave the result of a simple benchmark experiment.

<details>
<summary>benchmark.py</summary>
<pre>
<code>
import optuna
import numpy as np


optuna.logging.set_verbosity(optuna.logging.ERROR)

def objective(trial):
    c = trial.suggest_categorical("c", ["A", "B"])
    dim = 5

    if c == "A":
        x = np.zeros(dim)
        for i in range(dim):
            x[i] = trial.suggest_float(f"x{i}", -3, 3)
        return np.exp(-np.linalg.norm(x - 1)-np.linalg.norm(x+1))
    else:
        y = np.zeros(dim)
        for i in range(dim):
            y[i] = trial.suggest_float(f"y{i}", -10, 10)
        return np.exp(-np.linalg.norm(y))

n_trials = 300

study1_best_values = []
study2_best_values = []
study3_best_values = []

for i in range(100):
    sampler = optuna.samplers.TPESampler(multivariate=True, group=True)
    study1 = optuna.create_study(sampler=sampler, direction="maximize")
    study1.optimize(objective, n_trials=n_trials)

    sampler = optuna.samplers.TPESampler(multivariate=True)
    study2 = optuna.create_study(sampler=sampler, direction="maximize")
    study2.optimize(objective, n_trials=n_trials)

    sampler = optuna.samplers.TPESampler()
    study3 = optuna.create_study(sampler=sampler, direction="maximize")
    study3.optimize(objective, n_trials=n_trials)
    
    study1_best_values.append(study1.best_value)
    study2_best_values.append(study2.best_value)
    study3_best_values.append(study3.best_value)

print(f"Best value: {np.mean(study1_best_values)} +/- {np.std(study1_best_values)} with multivariate TPE (this PR)")
print(f"Best value: {np.mean(study2_best_values)} +/- {np.std(study2_best_values)} with multivariate TPE (master)")
print(f"Best value: {np.mean(study3_best_values)} +/- {np.std(study3_best_values)} with independent TPE")
</code>
</pre>
</details>

Results:
```bash
(venv) mamu@HideakinoMacBook-puro 2526 % python benchmark.py
Best value: 0.06591802692046024 +/- 0.11552717062807627 with multivariate TPE (this PR)
Best value: 0.033489006400232724 +/- 0.10152254960831676 with multivariate TPE (master)
Best value: 0.038333983591873635 +/- 0.11777197583658063 with independent TPE
```

hvy commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -192,6 +258,18 @@ def __init__(
+        if group:
+            if not multivariate:
+                raise ValueError(
+                    "``group`` option can only be enabled when ``multivariate`` is " "enabled."
```
```suggestion
                    "``group`` option can only be enabled when ``multivariate`` is enabled."
```

hvy commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -192,6 +258,18 @@ def __init__(
+                    "``group`` option can only be enabled when ``multivariate`` is " "enabled."
+                )
+            else:
+                warnings.warn(
```
Not a strong opinion but I think you can raise this warning independent of `multivariate`.

hvy commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -816,6 +914,8 @@ def _get_multivariate_observation_pairs(
     scores = []
     values: Dict[str, List[Optional[float]]] = {param_name: [] for param_name in param_names}
     for trial in study.get_trials(deepcopy=False, states=(TrialState.COMPLETE, TrialState.PRUNED)):
+        if any([param_name not in trial.params for param_name in param_names]):
+            continue
```
Might be helpful to leave a comment for future maintainers to why this condition is needed since it's not clear at first sight.

hvy commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -204,6 +282,13 @@ def infer_relative_search_space(
+        if self._group:
+            self._search_space_group = _GroupDecomposedSearchSpace(True).calculate(study)
+            _search_space = {}
+            for sub_space in self._search_space_group.group:
+                _search_space.update(sub_space)
+            return _search_space
```
For the intersection search space, we check that the distribution is included in `_DISTRIBUTION_CLASSES`. Should we make the same check here?

HideakiImamura replied:
We don't currently have any plans to add a distribution class that cannot be sampled in TPE, so I will just leave a comment here to simplify the implementation. What do you think?

HideakiImamura commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,96 @@
+        self._group: List[Dict[str, BaseDistribution]] = []
+
+    @property
+    def group(self) -> List[Dict[str, BaseDistribution]]:
```
Sounds good. I'll change `group` to `search_spaces`.

HideakiImamura commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -204,34 +283,62 @@ def infer_relative_search_space(
-            "to `False` in the constructor of `TPESampler`, "
-            "if this independent sampling is intended behavior.".format(param_name, trial.number)
-        )
+    def _log_independent_sampling(self, study: Study, trial: FrozenTrial, param_name: str) -> None:
+        if self._warn_independent_sampling:
+            complete_trials = study.get_trials(deepcopy=False)
+            if len(complete_trials) >= self._n_startup_trials:
+                _logger.warning(
+                    "The parameter '{param_name}' in trial#{trial.number} is sampled "
+                    "independently instead of being sampled by multivariate TPE sampler. "
+                    "(optimization performance may be degraded). "
+                    "You can suppress this warning by setting `warn_independent_sampling` "
+                    "to `False` in the constructor of `TPESampler`, "
+                    "if this independent sampling is intended behavior."
+                )
```
I changed the logic of this method to reduce redundancy. This change is reasonable because `log_independent_sampling` is called only when `_warn_independent_sampling` is `True`.

hvy replied:
Note that you can cache the trial count. It's not that critical since this method isn't called in the normal case.

hvy commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,100 @@
+            states_of_interest = (TrialState.COMPLETE,)
+
+        next_cursor = self._cursor
+        trials = study.get_trials(deepcopy=False, states=states_of_interest)
```
I think you want to go through all trials (incl. running ones) to correctly position the cursor.

hvy reviewed (commented):
Thanks for the quick fix. I left another comment now on the cursor logic if you could take a look.

not522 commented:
The cursor pattern may also be useful in other codes (e.g., `_pop_waiting_trial_id`). https://github.com/optuna/optuna/blob/1b2a391e6dbf6dfa3b06387b4a1d886fde4da911/optuna/study.py#L971-L984 How about split the method into the part independent of the class and implementation of actual processing?

hvy commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -204,34 +283,64 @@ def infer_relative_search_space(
+            assert self._search_space_group is not None
+            params = {}
+            for sub_space in self._search_space_group.search_spaces:
+                for name, distribution in sub_space.items():
+                    if not isinstance(distribution, _DISTRIBUTION_CLASSES):
+                        del sub_space[name]
```
You cannot change the size of the container while iterating over it. Let's create a new/copy dictionary of `sub_space`.

hvy reviewed (changes requested):
I left a single comment but I think it LGTM otherwise.

HideakiImamura commented:
@hvy Thanks for the insightful reviews. I addressed all of your reviews. PTAL.

@not522 Thanks for the comment. The cursor pattern is actually useful in several places, so there is some value in generalizing it. However, I think it is beyond the scope of this PR and is better considered as a follow-up.

hvy reviewed (approved):
LGTM!

I found some checks in the sampler tests `assert study.trials[-1].distributions ...` a bit superfluous but it's a minor.

Supporting this in other relative samplers such as `CmaEsSampler` in the future could benefit many users too that'd otherwise fall back to random sampling by default.

not522 reviewed (commented):
I found other bug in `_GroupDecomposedSearchSpace`. Please fix it.

I feel this process is complicated. We can add distributions for all trials in this PR and will implement the effective approach as a follow-up.

not522 commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,103 @@
+        if keys.isdisjoint(dist_keys):
+            continue
+
+        if keys < dist_keys:
+            return _add_distributions(
+                search_spaces, {name: distributions[name] for name in dist_keys - keys}
+            )
+
+        if keys > dist_keys:
+            search_spaces.append(distributions)
+            search_spaces.append({name: search_space[name] for name in keys - dist_keys})
+            search_spaces.remove(search_space)
+            return search_spaces
+
+        intersection = keys & dist_keys
+        search_spaces.append({name: search_space[name] for name in intersection})
+        if len(keys - intersection) > 0:
+            search_spaces.append({name: search_space[name] for name in keys - intersection})
+        search_spaces.remove(search_space)
+        return _add_distributions(
+            search_spaces, {name: distributions[name] for name in dist_keys - intersection}
+        )
```
This logic can be more straightforward by removing the if statements.
```suggestion
        intersection = keys & dist_keys
        new_search_spaces = []
        new_search_spaces.append({name: search_space[name] for name in intersection})
        new_search_spaces.append({name: search_space[name] for name in keys - intersection})
        new_search_spaces.append({name: distributions[name] for name in dist_keys - intersection})
        search_spaces.remove(search_space)
        for new_search_space in new_search_spaces:
            search_spaces = _add_distributions(search_spaces, new_search_space)

        return search_spaces
```

not522 replied:
Or we can rewrite this function as non-recursive.

HideakiImamura replied:
Thanks for the suggestion. I'll work on the refactoring as a follow-up as we discussed privately.

not522 commented on optuna/samplers/_search_space/group_decomposed.py:
```diff
@@ -0,0 +1,103 @@
+
+
+def _add_distributions(
+    search_spaces: List[Dict[str, BaseDistribution]], distributions: Dict[str, BaseDistribution]
```
Would you like to add a comment that `search_spaces` should be disjoint?

not522 commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -147,10 +149,71 @@ def objective(trial):
+            If this and ``multivariate`` are :obj:`True`, the multivariate TPE with the group
+            decomposed search space is used when suggesting parameters.
+            The sampling algorithm decomposes the search space based on past trials and samples
+            from the joint distribution in each decomposed subspace.
```
How about adding the description for the mathematical property of decomposed subspaces? For example, "The decomposed subspaces are a partition of the parameter set. Each subspace is a maximal subset subject to any search spaces should be superset or disjoint set of the subspace."

HideakiImamura commented:
@not522 Thanks for the reviews. I've decided to remove the logic of "cursor" since now the logic is not sufficiently tested. I'd like to work on that as a follow-up.

not522 commented on optuna/samplers/_tpe/sampler.py:
```diff
@@ -147,10 +149,74 @@ def objective(trial):
+            decomposed search space is used when suggesting parameters.
+            The sampling algorithm decomposes the search space based on past trials and samples
+            from the joint distribution in each decomposed subspace.
+            The decomposed subspaces are a partition of the whole search space. Each subspace is a
+            maximal subset of the entire search space under the condition that they are disjoint
+            from each other and that a trial exists whose parameter set includes the subspace.
```
Maybe it is incorrect. For example,

- search spaces are {x, y} and {y, z}
- subspaces are {x, y} and {z}

This case satisfies the description, but subspaces should be {x}, {y}, and {z}.

HideakiImamura replied:
:tashikani: Thank you for the comment. I updated the description.

not522 replied:
Could you remain the word "maximal"? The following case satisfies the definition, but the subspace should be {x, y}.

- search space is {x, y}
- subspaces are {x} and {y}

not522 reviewed (approved):
LGTM!

hvy commented on tests/samplers_tests/search_space_tests/test_group_decomposed.py:
```diff
@@ -0,0 +1,218 @@
+    assert search_space.calculate(study).search_spaces == []
+
+    # If two parameters have the same name but different distributions,
+    # the second one takes priority.
```
A huge detail. Unless we revert the logic (to reverse iterate).
```suggestion
    # the first one takes priority.
```

hvy reviewed (approved):
LGTM! 

Thanks @not522 for the careful review too. I agree that we can discuss the cursor in a separate ticket.
